### DISCUSSION TITLE
Showing the tool we implemented using Langium: A DSL to describe datasets for ML

### DISCUSSION THREAD

**Discussion by JoanGi** [eclipse-langium/langium#710] [Show and tell]
Following the threads in this section, I think it will be a good point to show the work we have done using Langium. I hope this could be useful for anyone who is starting to use this great toolkit to create DSLs.

The tool we developed with Langium is called [DescribeML](https://github.com/SOM-Research/DescribeML), which implement a DSL to describe dataset for machine-learning. The prototype will be presented in the [MODELS 2022 conference](https://conf.researchr.org/details/models-2022/models-2022-tools---demonstrations/5/DescribeML-a-tool-for-describing-machine-learning-datasets) and you can take a look at the paper [here](https://www.researchgate.net/publication/363256430_DescribeML_A_Tool_for_Describing_Machine_Learning_Datasets) and at a video presentation [here](https://www.youtube.com/watch?v=Bf3bhWB-UJY). 

If you like it, please, support us with a star in the repo ;)

## The DSL goal:

The goal was to implement a declarative DSL of a set of concepts we detected as relevant to describe datasets intended to train machine learning (ML) models. For instance, an image dataset with few dark faces could add a racial bias in a face classifier, so better know this issues before using this data to train your ML model.

An example of a grammar instance:

```ruby
Dataset: Melanoma_Classification_Dataset 
    Metadata: 
        Title: "2020 SIIM-ISIC Melanoma Classification Challenge Dataset" 
        Unique-identifier: SIC_Melanoma_Classification_Challenge_Dataset
        Version: v0001 
        Dates:
            Release Date: 10-08-20
        Citation: 
            Title: "A patient-centric dataset of images and metadata for identifying melanomas using clinical context"
            Authors: "Rotemberg, V., ..."
            Year: 2021 
            Journal: "Sci Data 8, 34"
            DOI: 10.1038/s41597-021-00815-z
        Description:
            Purposes: "The 2020 SIIM-ISIC Melanoma Classification challenge dataset..." 
            Tasks:    [classification, image-classification]
            Gaps:     "As the leading healthcare organization for informatics in medical imaging..."  
            Areas:    HealthCare
            Tags:     Images Melanoma 
        Applications:
            Past Uses: "Yes. The 2020 SIIM-ISIC Melanoma Classification ..."
            Recommended: 
                "Identify melanoma in lesion images" 
                "Predict incidence of melanoma in a population"
            Non-recommended: "Due to low population..."
        Distribution:
            Licences: CC BY 3.0 (Attribution 3.0 Unported)
            Rights(stand-alone): Access
            Rights(with models): Benchmark
            Additional Conditions "In addition to the CC-BY-NC license, the dataset is governed by the ISIC Terms of Use. Learn more about the terms of use here: https://datanutrition.org/labels/isic-2020/ "                          
    Authoring:
        Authors:
            Name "Skin Imaging Collaboration ISIC"  email emailo@emailo.com
        Funders:
            Name "The_University_of_Queensland"  type mixed 
                grantor "National Health and Medical Research Council (NHMRC) – Centre of Research Excellence Scheme" 
                grantId: APP1099021
            Erratum?: "There is no erratum known" 
            Contribution guidelines: "No contribution guidelines provided"
```

In addition to the grammar, we have implemented: 

- A custom validator service to provide hints, during the creation of the description files.
- An uploader services that allow users to upload .CSV files and creates a semi-completed draft of a description from a dataset.
- A generator service that get a valid description and generates HTML documents.

## Which troubles are we facing?

- I'm trying to guess how I can force the indentation of every term.  (similar to Python).
- The grammar declaration are becoming too big, I'm are looking for a way to split the grammar declaration in more files (so it can be easily maintainable).
- Every time we generate the code from a grammar, the syntax is generated (in our case we use it for syntactic highlight). If you want to implement semantic highlight, these are deleted every time. We are looking how we can solve it. (I'm sure, there is an already built-in solution, just we didn't found it yet).
- I'm sure there are more... I will write it on the comments as soon they come to my mind.

## Which are our plans?

- Fix the mentioned troubles
- Migrate from the 0.0.2 to the new 0.0.5 release.
- Implement the DSL a stand-alone web component to embed in any website. Like the [machine-state showcase.](https://langium.org/showcase/statemachine/)

I hope this example of a declarative DSL will help anyone, and if you have a question or an answer to our troubles, please ask here, and we will learn together.

At last, thanks again to the Langium team for this incredible work 👍

**Comment by snarkipus** [comment]
>RE: Every time we generate the code from a grammar, the syntax is generated (in our case we use it for syntactic highlight). If you want to implement semantic highlight, these are deleted every time. We are looking how we can solve it. (I'm sure, there is an already built-in solution, just we didn't found it yet).

There was  post recently where someone had a similar issue which was resolved by merging the textmate grammar in a separate step ([link](https://github.com/langium/langium/discussions/604#discussioncomment-3183135)). @msujew might have a more clever way though.

**Reply by spoenemann** [reply]
The easiest solution is to remove the `textMate` entry from langium-config.json, this will stop regenerating the syntax file and you can fine-tune it manually instead.

**Reply by JoanGi** [reply]
That's it! Thanlks :)!!!

**Comment by spoenemann** [comment]
Hi @JoanGi, thanks for presenting your tool!

> The grammar declaration are becoming too big, I'm are looking for a way to split the grammar declaration in more files (so it can be easily maintainable).

You can import another grammar file to access all its declaration. An example is available here:
https://github.com/langium/langium/blob/main/examples/requirements/src/language-server/tests.langium
